Add a line search in newton solve to find contact point

GiudGiud opened the pull request:
Dont die horribly when it fails
refs #28148
See https://github.com/idaholab/moose/discussions/28091 and https://github.com/idaholab/moose/discussions/27934

Note that the line search does not save this input. We still dont converge to the right contact point, and get a warning along with debugging information from libmesh.

GiudGiud committed:
Add a line search to try to solve contact problems
Dont die horribly when it fails
refs #28148

GiudGiud committed:
Make sure output of newton search is always in a decent state
Handle problematic case

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2330746/) on f54e6a5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28149/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2330761/) on f54e6a5 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>5bf0d8</th>
      <th colspan=4>#28149 f54e6a</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.06%</td>
      <td>85.05%</td>
      <td>-0.01%</td>
      <td>51.61%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104544</td>
      <td>104549</td>
      <td>+5</td>
      <td>16</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18360</td>
      <td>18377</td>
      <td>+17</td>
      <td>15</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28149/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28149/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28149/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 51.61% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
so I looked at this more after talking with @bwspenc 

we already check that we find a face for every secondary node, so these changes are just weeding out bad faces (previously accepted by not checking the newton tolerance...)

Thoughts on the way forward here? I have a 600kb mesh from discussion we could use for a test case. I could probably get it down to 300kb

GiudGiud commented:
why is the checking on not having leftover unmatched secondary nodes only done on Jacobian computation?
```
  if (recheck_secondary_nodes.size() > 0 && _patch_update_strategy != Moose::Iteration &&
      _subproblem.currentlyComputingJacobian())
    mooseDoOnce(mooseWarning("Warning in PenetrationLocator. Penetration is not "
```

GiudGiud committed:
Avoid adding contact points penetration info
Small optimizations in search solves
Dont include verbose output for failed searches, we do not need all contact point searches to pass, only one and only the best one

GiudGiud commented:
ok now there's real failures from rejecting nodes

lindsayad commented:
this warning was added in https://github.com/idaholab/moose/pull/9832

GiudGiud commented:
@sveerara why only on Jacobian?

GiudGiud committed:
Undo optimizations which cause a diff
Add variables to avoid being inconsistent in all the heuristics

GiudGiud committed:
Add a test based on the case that was causing the problem
mesh is reduced to only the problematic area

lindsayad commented on framework/src/geomsearch/FindContactPoint.C:
```diff
@@ -171,32 +177,74 @@ findContactPoint(PenetrationInfo & p_info,
+    // Improvised line search in case the update is too large and gets out of the element so bad
+    // that we cannot reinit at the new point
+    Real mult = 1;
+    bool success = false;
+    while (!success)
```
```suggestion
    while (true)
```

lindsayad replied:
If you don't want to incorporate a suggestion, please give a rationale instead of just marking it as resolved

GiudGiud replied:
pretty sure I applied the comments lemme look

GiudGiud replied:
redone. I must have force-pushed it off the branch

lindsayad commented on framework/src/geomsearch/FindContactPoint.C:
```diff
@@ -171,32 +177,74 @@ findContactPoint(PenetrationInfo & p_info,
-    d = secondary_point - phys_point[0];
+        points[0] = ref_point;
+        fe_side->reinit(side, &points);
+        success = true;
```
```suggestion
```

lindsayad commented on framework/src/geomsearch/FindContactPoint.C:
```diff
@@ -171,32 +177,74 @@ findContactPoint(PenetrationInfo & p_info,
 
-    update_size = update.l2_norm();
+        // we don't multiply by 'mult' because it is used for convergence
+        update_size = update.l2_norm();
```
```suggestion
        update_size = update.l2_norm();
        break;
```

lindsayad commented on framework/src/geomsearch/FindContactPoint.C:
```diff
@@ -171,32 +177,74 @@ findContactPoint(PenetrationInfo & p_info,
+            mooseWarning("We could not solve for the contact point.", e.what());
+          update_size = update.l2_norm();
+          d = (secondary_point - phys_point[0]) * mult;
+          success = true;
```
```suggestion
          break;
```

lindsayad commented on framework/src/geomsearch/FindContactPoint.C:
```diff
@@ -49,8 +52,13 @@ findContactPoint(PenetrationInfo & p_info,
+                 bool verbose)
 {
+  // Default to true and we'll switch on failures
+  search_succeeded = true;
```
AFAICT you never switch this to `false`?

GiudGiud replied:
added now

lindsayad commented on framework/src/geomsearch/FindContactPoint.C:
```diff
@@ -205,7 +253,8 @@ findContactPoint(PenetrationInfo & p_info,
   {
     p_info._normal = dxyz_dxi[0].cross(dxyz_deta[0]);
-    p_info._normal /= p_info._normal.norm();
+    if (!MooseUtils::absoluteFuzzyEqual(p_info._normal.norm(), 0))
```
The normal can be 0?

GiudGiud replied:
yeah seen . see in 3d there's the same

GiudGiud replied:
right below

lindsayad commented on framework/include/geomsearch/FindContactPoint.h:
```diff
@@ -34,7 +34,9 @@ void findContactPoint(PenetrationInfo & p_info,
-                      bool & contact_point_on_side);
+                      bool & contact_point_on_side,
+                      bool & search_succeeded,
+                      bool verbose = false);
```
This parameter seems to not be used

GiudGiud replied:
removed

GiudGiud commented:
Ok forgot to commit some code
My bad

GiudGiud commented on framework/src/geomsearch/FindContactPoint.C:
```diff
@@ -171,32 +175,75 @@ findContactPoint(PenetrationInfo & p_info,
+        if (mult < 1e-6)
+        {
+#ifndef NDEBUG
+          mooseWarning("We could not solve for the contact point.", e.what());
```
using the macro in lieu of 'verbose' since I would have to add verbose booleans to 3 classes to propagate the option

GiudGiud committed:
Address Alex review

better while loop format

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented:
Failures unrelated

roystgnr commented on framework/src/geomsearch/PenetrationThread.C:
```diff
@@ -1723,18 +1727,23 @@ PenetrationThread::createInfoForElem(std::vector<PenetrationInfo *> & thisElemIn
+    {
+      thisElemInfo.push_back(pen_info);
+      p_info.push_back(pen_info);
+    }
```
```
else
  break_valgrind_tests_with(MEMORY_LEAK);
```

I'll put up a fix shortly.

GiudGiud replied:
thanks
